# Prédiction Match Depuis Les Clusters Joueurs

Objectif : mesurer si une composition de styles de joueurs est plus forte qu'une autre pour prédire le résultat d'un match (`1`, `0`, `-1`).

Le notebook compare deux représentations :

- baseline simple : les `cluster_key` des 22 titulaires en colonnes ;
- cross-features : des comptes de styles par équipe et des différences home-away.

La feature ML reste toujours `cluster_key`, jamais `cluster_id` seul.


## TL;DR Des Résultats Actuels

La baseline en `22 slots` surapprend fortement : le RandomForest global passe de `0.643` de balanced accuracy train à `0.399` en test, et le RandomForest same bucket passe de `0.688` à `0.393`.

La nouvelle validation temporelle régularisée réduit nettement ce problème, mais elle montre aussi une conclusion importante : les styles seuls ne suffisent pas encore.

- Global `styles_only` : test balanced accuracy `0.374`, gap `0.128`.
- Global `ranking_only` : test balanced accuracy `0.429`, gap `0.002`.
- Global `styles_plus_ranking` : test balanced accuracy `0.435`, gap `0.037`.
- Same bucket `styles_only` : test balanced accuracy `0.360`, gap `0.176`.
- Same bucket `ranking_only` : test balanced accuracy `0.450`, gap `-0.021`.
- Same bucket `styles_plus_ranking` : test balanced accuracy `0.428`, gap `0.094`.

Conclusion data scientist : le signal principal vient encore du niveau relatif des clubs. Les compositions de styles apportent peut-être un petit gain en global, mais ce gain n'est pas robuste sur le filtre same bucket. Pour vraiment mesurer la force d'une composition de styles, il faudra réduire la dimension des styles ou construire des styles collectifs plus stables.


## 1. Setup

Paramètres, imports et chemins utilisés dans le notebook.

In [78]:
from pathlib import Path
import os
import re
import unicodedata

import numpy as np
import pandas as pd
from IPython.display import display
from sqlalchemy import create_engine

pd.set_option("display.max_columns", 160)
pd.set_option("display.max_rows", 120)

PROJECT_ROOT = Path.cwd()
if PROJECT_ROOT.name == "notebooks":
    PROJECT_ROOT = PROJECT_ROOT.parent

SCHEMA = "public"
RANDOM_STATE = 42
TEST_SIZE = 0.20
RANKING_YEAR = 2026
TARGET_LABELS = [-1, 0, 1]

CLUSTERS_PATH = PROJECT_ROOT / "analysis_outputs" / "player_role_clusters_with_style_df.csv"
RANKING_PATH = Path(
    "/Users/hugopierre/projets/formation jedha/foot-predictor/data/raw/club_clustering/uefa_top_100_clubs.csv"
)
OUTPUT_DIR = PROJECT_ROOT / "analysis_outputs"
OUTPUT_DIR.mkdir(exist_ok=True)

POSITION_ORDER = {
    "Goalkeeper": 1,
    "Centre-Back": 2,
    "Left-Back": 3,
    "Right-Back": 4,
    "Sweeper": 5,
    "Defensive Midfield": 6,
    "Central Midfield": 7,
    "Attacking Midfield": 8,
    "Left Midfield": 9,
    "Right Midfield": 10,
    "Left Winger": 11,
    "Right Winger": 12,
    "Second Striker": 13,
    "Centre-Forward": 14,
}

print("Projet:", PROJECT_ROOT)
print("Clusters:", CLUSTERS_PATH)
print("Ranking:", RANKING_PATH)


Projet: /Users/hugopierre/projets/formation jedha/football_predictor/football-predictor
Clusters: /Users/hugopierre/projets/formation jedha/football_predictor/football-predictor/analysis_outputs/player_role_clusters_with_style_df.csv
Ranking: /Users/hugopierre/projets/formation jedha/foot-predictor/data/raw/club_clustering/uefa_top_100_clubs.csv


## 2. Connexion Neon

La connexion lit `NEON_DATABASE_URL`, `DATABASE_URL` ou `POSTGRES_URL` depuis l'environnement ou le `.env`.

In [79]:
def load_env_file(path: Path) -> dict[str, str]:
    values = {}

    if not path.exists():
        return values

    for raw_line in path.read_text(encoding="utf-8").splitlines():
        line = raw_line.strip()

        if not line or line.startswith("#") or "=" not in line:
            continue

        key, value = line.split("=", 1)
        values[key.strip()] = value.strip().strip('"').strip("'")

    return values


def sqlalchemy_url(url: str) -> str:
    if url.startswith("postgres://"):
        return url.replace("postgres://", "postgresql+psycopg2://", 1)

    if url.startswith("postgresql://"):
        return url.replace("postgresql://", "postgresql+psycopg2://", 1)

    return url


env_path = PROJECT_ROOT / ".env"
local_env = load_env_file(env_path)
DATABASE_URL = (
    os.environ.get("NEON_DATABASE_URL")
    or os.environ.get("DATABASE_URL")
    or os.environ.get("POSTGRES_URL")
    or local_env.get("NEON_DATABASE_URL")
    or local_env.get("DATABASE_URL")
    or local_env.get("POSTGRES_URL")
    or ""
)

if not DATABASE_URL:
    raise RuntimeError("Ajoute NEON_DATABASE_URL, DATABASE_URL ou POSTGRES_URL dans .env.")

engine = create_engine(sqlalchemy_url(DATABASE_URL), pool_pre_ping=True)
print("Connexion configurée")

Connexion configurée


## 3. Charger Les Données

On charge les tables nécessaires depuis Neon. Pour accélérer, `lineup` est simplement limité aux titulaires.


In [80]:
matches_df = pd.read_sql(
    f'''
    SELECT match_id, match_date, competition_id, round, stadium, referee
    FROM "{SCHEMA}"."match"
    ''',
    engine,
)

match_team_df = pd.read_sql(
    f'''
    SELECT match_id, team_id, side, score, penalty_score, manager_name, formation
    FROM "{SCHEMA}"."match_team"
    ''',
    engine,
)

lineup_df = pd.read_sql(
    f'''
    SELECT match_id, team_id, player_id, position_player, is_starting_match
    FROM "{SCHEMA}"."lineup"
    WHERE is_starting_match = 1
    ''',
    engine,
)

teams_df = pd.read_sql(
    f'''
    SELECT team_id, team_name
    FROM "{SCHEMA}"."team"
    ''',
    engine,
)

clusters_df = pd.read_csv(CLUSTERS_PATH)
ranking_raw_df = pd.read_csv(RANKING_PATH)

print("match", matches_df.shape)
print("match_team", match_team_df.shape)
print("lineup titulaires", lineup_df.shape)
print("team", teams_df.shape)
print("clusters", clusters_df.shape)
print("ranking", ranking_raw_df.shape)

display(matches_df.head())
display(match_team_df.head())
display(lineup_df.head())
display(clusters_df.head())


match (4052, 6)
match_team (8104, 7)
lineup titulaires (89144, 5)
team (99, 2)
clusters (5046, 11)
ranking (100, 11)


,match_id,match_date,competition_id,round,stadium,referee
0,3583624,2021-07-07,CLQ,First Round 1st leg,Aspmyra Stadion,Juxhin Xhaja
1,3583639,2021-07-14,CLQ,First Round 2nd leg,Stadion Miejski im. Marszałka Józefa Piłsudskiego,Aristotelis Diamantopoulos
2,3604605,2021-07-17,BESC,Final,Jan Breydelstadion,Jonathan Lardot
3,3584558,2021-07-20,CLQ,Second Round 1st leg,Celtic Park,Sandro Schärer
4,3615079,2021-07-20,CLQ,Second Round 1st leg,Maksimir,Pawel Gil


,match_id,team_id,side,score,penalty_score,manager_name,formation
0,3573730,583,away,0,None,Mauricio Pochettino,4-3-3 Attacking
1,3573730,1082,home,1,None,Jocelyn Gourvennec,4-4-2 double 6
2,3575344,27,away,3,None,Julian Nagelsmann,4-2-3-1
3,3575344,16,home,1,None,Marco Rose,4-4-2 Diamond
4,3578075,281,away,0,None,Pep Guardiola,4-3-3 Attacking


,match_id,team_id,player_id,position_player,is_starting_match
0,3573730,583,99343,Central Midfield,1
1,3573730,583,536482,Central Midfield,1
2,3573730,583,282041,Centre-Back,1
3,3573730,583,228948,Centre-Back,1
4,3573730,583,68863,Centre-Forward,1


,player_id,player_name,best_position,position_role,cluster_id,cluster_key,style_label,style_strengths,style_weaknesses,distance_to_centroid,chosen_k
0,652061,Adam Karabec,CAM,attacking_midfielder,0,attacking_midfielder_0,milieu offensif - coup franc / centre / effet,"coup franc (+0.90), centre / effet (+0.85), pe...","accélération / vitesse (-0.96), endurance (-0....",3.295,3
1,43530,Adam Lallana,CAM,attacking_midfielder,0,attacking_midfielder_0,milieu offensif - coup franc / centre / effet,"coup franc (+0.90), centre / effet (+0.85), pe...","accélération / vitesse (-0.96), endurance (-0....",3.959,3
2,45287,Adel Taarabt,CAM,attacking_midfielder,0,attacking_midfielder_0,milieu offensif - coup franc / centre / effet,"coup franc (+0.90), centre / effet (+0.85), pe...","accélération / vitesse (-0.96), endurance (-0....",4.940,3
3,255450,Aleksey Miranchuk,CAM,attacking_midfielder,0,attacking_midfielder_0,milieu offensif - coup franc / centre / effet,"coup franc (+0.90), centre / effet (+0.85), pe...","accélération / vitesse (-0.96), endurance (-0....",3.789,3
4,130365,Alessandro Florenzi,CAM,attacking_midfielder,0,attacking_midfielder_0,milieu offensif - coup franc / centre / effet,"coup franc (+0.90), centre / effet (+0.85), pe...","accélération / vitesse (-0.96), endurance (-0....",3.595,3


## 4. Contrôler Les Clusters Joueurs

On vérifie que la feature ML est bien `cluster_key`, unique par rôle + cluster, et jamais `cluster_id` seul.

In [81]:
required_cluster_columns = [
    "player_id", "best_position", "position_role", "cluster_id", "cluster_key", "style_label"
]
missing_cluster_columns = [column for column in required_cluster_columns if column not in clusters_df.columns]
if missing_cluster_columns:
    raise KeyError(f"Colonnes manquantes dans le CSV clusters: {missing_cluster_columns}")

clusters_df = clusters_df[required_cluster_columns].copy()
clusters_df["player_id"] = pd.to_numeric(clusters_df["player_id"], errors="raise").astype(int)
clusters_df["cluster_id"] = pd.to_numeric(clusters_df["cluster_id"], errors="raise").astype(int)
clusters_df["expected_cluster_key"] = clusters_df["position_role"] + "_" + clusters_df["cluster_id"].astype(str)

bad_cluster_key_df = clusters_df[clusters_df["cluster_key"] != clusters_df["expected_cluster_key"]]
if not bad_cluster_key_df.empty:
    raise ValueError("Certains cluster_key ne correspondent pas à position_role + cluster_id.")

cluster_key_control_df = (
    clusters_df
    .groupby("cluster_key")
    .agg(
        position_role_count=("position_role", "nunique"),
        cluster_id_count=("cluster_id", "nunique"),
        player_count=("player_id", "count"),
        style_label=("style_label", "first"),
    )
    .reset_index()
    .sort_values("cluster_key")
)

if cluster_key_control_df["position_role_count"].max() != 1:
    raise ValueError("Un cluster_key correspond à plusieurs rôles.")

if cluster_key_control_df["cluster_id_count"].max() != 1:
    raise ValueError("Un cluster_key correspond à plusieurs cluster_id.")

if cluster_key_control_df["cluster_key"].nunique() != 29:
    raise ValueError("On attend exactement 29 cluster_key distincts.")

clusters_df = clusters_df.drop(columns="expected_cluster_key")

print("Cluster_key OK:", cluster_key_control_df["cluster_key"].nunique(), "clusters distincts")
display(cluster_key_control_df)

Cluster_key OK: 29 clusters distincts


,cluster_key,position_role_count,cluster_id_count,player_count,style_label
0,attacking_midfielder_0,1,1,166,milieu offensif - coup franc / centre / effet
1,attacking_midfielder_1,1,1,259,milieu offensif - agilité / équilibre / accélé...
2,attacking_midfielder_2,1,1,339,milieu offensif - conscience défensive / inter...
3,central_defender_0,1,1,100,défenseur central - gestuelle / placement offe...
4,central_defender_1,1,1,231,défenseur central - calme / réactivité / consc...
5,central_defender_2,1,1,350,défenseur central - tête / détente / accélérat...
6,central_defender_3,1,1,317,défenseur central - centre / effet / coup franc
7,central_midfielder_0,1,1,62,milieu central - centre / effet / coup franc
8,central_midfielder_1,1,1,119,milieu central - conscience défensive / interc...
9,central_midfielder_2,1,1,110,milieu central - tête / détente / force


## 5. Aperçu Joueur -> Cluster

Petit aperçu de 5 joueurs avec le `cluster_key` correspondant.


In [82]:
players_head_df = pd.read_sql(
    f'''
    SELECT player_id, player_name, best_position
    FROM "{SCHEMA}"."player"
    ORDER BY player_name
    LIMIT 100
    ''',
    engine,
)

players_head_df["player_id"] = pd.to_numeric(players_head_df["player_id"], errors="raise").astype(int)

players_head_df = (
    players_head_df
    .merge(clusters_df[["player_id", "cluster_key"]], on="player_id", how="inner")
    [["player_id", "player_name", "best_position", "cluster_key"]]
    .head(5)
)

display(players_head_df)


,player_id,player_name,best_position,cluster_key
0,889193,AZ Jackson,CAM,attacking_midfielder_1
1,434207,Aaron Connolly,ST,forward_2
2,92571,Aaron Cresswell,LB,fullback_wingback_2
3,591949,Aaron Hickey,RB,fullback_wingback_1
4,123951,Aaron Mooy,CM,central_midfielder_3


## 6. Préparer Le Ranking Top 100

Le ranking vient du CSV existant `uefa_top_100_clubs.csv`. Il sert uniquement à filtrer les matchs où les deux équipes sont dans le même bucket de 20 places.

In [83]:
def normalize_team_name(value):
    if pd.isna(value):
        return ""

    text = str(value).replace("ß", "ss")
    text = unicodedata.normalize("NFKD", text).encode("ascii", "ignore").decode("ascii")
    text = text.lower().replace("&", " and ")
    text = re.sub(r"[^a-z0-9]+", " ", text)
    text = re.sub(r"\s+", " ", text).strip()
    return text


TEAM_NAME_TO_RANKING_KEY_OVERRIDES = {
    "1 fc union berlin": "union berlin",
    "ac milan": "milan",
    "ac sparta prague": "sparta praha",
    "acf fiorentina": "fiorentina",
    "as monaco": "monaco",
    "arsenal fc": "arsenal",
    "associazione sportiva roma": "roma",
    "atalanta bc": "atalanta",
    "atletico de madrid": "atletico madrid",
    "bsc young boys": "young boys",
    "bayer 04 leverkusen": "bayer leverkusen",
    "bayern munich": "bayern munchen",
    "bologna football club 1909": "bologna",
    "celtic fc": "celtic",
    "chelsea fc": "chelsea",
    "club brugge kv": "club brugge",
    "djurgardens idrottsforening": "djurgardens if",
    "fc barcelona": "barcelona",
    "fc basel 1893": "basel",
    "fc copenhagen": "copenhagen",
    "fc lugano": "lugano",
    "fc midtjylland": "midtjylland",
    "fc porto": "porto",
    "fc shakhtar donetsk": "shakhtar donetsk",
    "fc viktoria plzen": "viktoria plzen",
    "fk partizan belgrade": "partizan",
    "ferencvarosi tc": "ferencvaros",
    "feyenoord rotterdam": "feyenoord",
    "fotballklubben bodo glimt": "bodo glimt",
    "fotballklubben bod glimt": "bodo glimt",
    "fussballclub red bull salzburg": "salzburg",
    "gnk dinamo zagreb": "dinamo zagreb",
    "inter milan": "inter",
    "juventus fc": "juventus",
    "kaa gent": "gent",
    "krc genk": "genk",
    "losc lille": "lille osc",
    "legia warszawa": "legia warsaw",
    "liverpool fc": "liverpool",
    "molde fotballklubb": "molde",
    "nk celje": "celje",
    "olympiakos syndesmos filathlon peiraios": "olympiakos",
    "olympique lyon": "lyon",
    "olympique marseille": "marseille",
    "pafos fc": "pafos",
    "panathinaikos athlitikos omilos": "panathinaikos",
    "panthessalonikios athlitikos omilos konstantinoupoliton": "paok thessaloniki",
    "qarabag fk": "qarabag",
    "rc strasbourg alsace": "strasbourg",
    "rsc anderlecht": "anderlecht",
    "rangers fc": "rangers",
    "rapid vienna": "rapid wien",
    "real betis balompie": "real betis",
    "sc braga": "sporting braga",
    "sc freiburg": "freiburg",
    "sk slavia prague": "slavia prague",
    "sl benfica": "benfica",
    "ssc napoli": "napoli",
    "sevilla fc": "sevilla",
    "societa sportiva lazio s p a": "lazio",
    "sporting cp": "sporting cp lisbon",
    "sportklub puntigamer sturm graz": "sturm graz",
    "stade rennais fc": "stade rennais",
    "vfb stuttgart": "stuttgart",
    "villarreal cf": "villarreal",
}

ranking_df = ranking_raw_df[ranking_raw_df["ranking_year"].eq(RANKING_YEAR)].copy()
ranking_df["ranking_key"] = ranking_df["club_key"].map(normalize_team_name)
ranking_df["rank_bucket"] = ((ranking_df["rank"] - 1) // 20 + 1).astype(int)

if ranking_df["ranking_key"].duplicated().any():
    duplicated_keys = ranking_df[ranking_df["ranking_key"].duplicated(keep=False)]
    raise ValueError(f"ranking_key dupliqués dans le ranking: {duplicated_keys[['club_name', 'ranking_key']]}")

ranking_lookup_df = ranking_df[[
    "ranking_year", "rank", "ranking_row", "club_name", "country_code",
    "total_points", "club_key", "ranking_key", "rank_bucket",
]].copy()

teams_rank_df = teams_df.copy()
teams_rank_df["team_normalized"] = teams_rank_df["team_name"].map(normalize_team_name)
teams_rank_df["ranking_key"] = teams_rank_df["team_normalized"].map(
    lambda value: TEAM_NAME_TO_RANKING_KEY_OVERRIDES.get(value, value)
)

team_rank_mapping_df = teams_rank_df.merge(
    ranking_lookup_df,
    on="ranking_key",
    how="left",
)

unmapped_teams_df = (
    team_rank_mapping_df[team_rank_mapping_df["rank"].isna()]
    [["team_id", "team_name", "team_normalized", "ranking_key"]]
    .sort_values("team_name")
    .reset_index(drop=True)
)

print("Équipes mappées ranking:", team_rank_mapping_df["rank"].notna().sum(), "/", len(team_rank_mapping_df))
print("Équipes non mappées:", len(unmapped_teams_df))
display(team_rank_mapping_df.sort_values("rank").head(20))
display(unmapped_teams_df)

Équipes mappées ranking: 98 / 99
Équipes non mappées: 1


,team_id,team_name,team_normalized,ranking_key,ranking_year,rank,ranking_row,club_name,country_code,total_points,club_key,rank_bucket
15,27,Bayern Munich,bayern munich,bayern munchen,2026.0,1.0,1.0,Bayern München,Ger,147.50,bayern munchen,1.0
80,418,Real Madrid,real madrid,real madrid,2026.0,2.0,2.0,Real Madrid,Esp,144.50,real madrid,1.0
70,583,Paris Saint-Germain,paris saint germain,paris saint germain,2026.0,3.0,3.0,Paris Saint-Germain,Fra,132.00,paris saint germain,1.0
52,31,Liverpool FC,liverpool fc,liverpool,2026.0,4.0,4.0,Liverpool,Eng,130.00,liverpool,1.0
42,46,Inter Milan,inter milan,inter,2026.0,5.0,5.0,Internazionale,Ita,127.00,inter,1.0
55,281,Manchester City,manchester city,manchester city,2026.0,6.0,6.0,Manchester City,Eng,125.50,manchester city,1.0
7,11,Arsenal FC,arsenal fc,arsenal,2026.0,7.0,7.0,Arsenal,Eng,119.00,arsenal,1.0
25,131,FC Barcelona,fc barcelona,barcelona,2026.0,8.0,8.0,FC Barcelona,Esp,113.25,barcelona,1.0
14,15,Bayer 04 Leverkusen,bayer 04 leverkusen,bayer leverkusen,2026.0,9.0,9.0,Bayer Leverkusen,Ger,105.00,bayer leverkusen,1.0
12,13,Atlético de Madrid,atletico de madrid,atletico madrid,2026.0,10.0,10.0,Atlético Madrid,Esp,104.75,atletico madrid,1.0


,team_id,team_name,team_normalized,ranking_key
0,10004,Paris FC,paris fc,paris fc


## 7. Construire Les 22 Features Match

Une ligne = un match. Les 11 titulaires de chaque équipe deviennent des colonnes catégorielles `cluster_key`.

In [84]:
home_team_df = (
    match_team_df[match_team_df["side"].eq("home")]
    .rename(columns={"team_id": "home_team_id", "score": "home_score"})
    [["match_id", "home_team_id", "home_score", "manager_name", "formation"]]
    .rename(columns={"manager_name": "home_manager_name", "formation": "home_formation"})
)

away_team_df = (
    match_team_df[match_team_df["side"].eq("away")]
    .rename(columns={"team_id": "away_team_id", "score": "away_score"})
    [["match_id", "away_team_id", "away_score", "manager_name", "formation"]]
    .rename(columns={"manager_name": "away_manager_name", "formation": "away_formation"})
)

base_match_df = (
    matches_df
    .merge(home_team_df, on="match_id", how="inner")
    .merge(away_team_df, on="match_id", how="inner")
    .merge(teams_df.rename(columns={"team_id": "home_team_id", "team_name": "home_team_name"}), on="home_team_id", how="left")
    .merge(teams_df.rename(columns={"team_id": "away_team_id", "team_name": "away_team_name"}), on="away_team_id", how="left")
)

base_match_df["home_score"] = pd.to_numeric(base_match_df["home_score"], errors="coerce")
base_match_df["away_score"] = pd.to_numeric(base_match_df["away_score"], errors="coerce")
base_match_df = base_match_df[base_match_df["home_score"].notna() & base_match_df["away_score"].notna()].copy()
base_match_df["target_result"] = 0
base_match_df.loc[base_match_df["home_score"] > base_match_df["away_score"], "target_result"] = 1
base_match_df.loc[base_match_df["home_score"] < base_match_df["away_score"], "target_result"] = -1

lineup_side_df = lineup_df.merge(
    match_team_df[["match_id", "team_id", "side"]],
    on=["match_id", "team_id"],
    how="left",
)

starters_df = lineup_side_df[lineup_side_df["is_starting_match"].fillna(0).astype(int).eq(1)].copy()
starters_df["player_id"] = pd.to_numeric(starters_df["player_id"], errors="raise").astype(int)
starters_df = starters_df.merge(
    clusters_df[["player_id", "cluster_key", "style_label"]],
    on="player_id",
    how="left",
)
starters_df["is_goalkeeper"] = starters_df["position_player"].eq("Goalkeeper")
starters_df["position_order"] = starters_df["position_player"].map(POSITION_ORDER).fillna(99).astype(int)
starters_df["has_cluster"] = starters_df["cluster_key"].notna()

team_lineup_quality_df = (
    starters_df
    .groupby(["match_id", "team_id", "side"], dropna=False)
    .agg(
        starter_count=("player_id", "count"),
        clustered_starter_count=("has_cluster", "sum"),
        goalkeeper_count=("is_goalkeeper", "sum"),
    )
    .reset_index()
)

complete_team_lineups_df = team_lineup_quality_df[
    team_lineup_quality_df["starter_count"].eq(11)
    & team_lineup_quality_df["clustered_starter_count"].eq(11)
    & team_lineup_quality_df["goalkeeper_count"].eq(1)
].copy()

complete_starters_df = starters_df.merge(
    complete_team_lineups_df[["match_id", "team_id", "side"]],
    on=["match_id", "team_id", "side"],
    how="inner",
)

complete_starters_df = complete_starters_df.sort_values([
    "match_id", "team_id", "side", "position_order", "cluster_key", "player_id"
]).copy()
complete_starters_df["slot_number"] = (
    complete_starters_df
    .groupby(["match_id", "team_id", "side"])
    .cumcount()
    .add(1)
)

bad_slot_df = complete_starters_df[complete_starters_df["slot_number"].gt(11)]
if not bad_slot_df.empty:
    raise ValueError("Certaines équipes ont plus de 11 slots après filtrage.")

first_slot_check_df = complete_starters_df[complete_starters_df["slot_number"].eq(1)]
if not first_slot_check_df["position_player"].eq("Goalkeeper").all():
    raise ValueError("Le slot 1 doit toujours être le gardien.")

slot_cluster_df = (
    complete_starters_df
    .pivot_table(
        index=["match_id", "team_id", "side"],
        columns="slot_number",
        values="cluster_key",
        aggfunc="first",
    )
    .reset_index()
)
slot_cluster_df.columns = [
    str(column) if not isinstance(column, int) else f"player_{column:02d}_cluster_key"
    for column in slot_cluster_df.columns
]

home_slots_df = (
    slot_cluster_df[slot_cluster_df["side"].eq("home")]
    .drop(columns="side")
    .rename(columns={"team_id": "home_team_id"})
)
home_slots_df = home_slots_df.rename(columns={
    f"player_{slot:02d}_cluster_key": f"home_player_{slot:02d}_cluster_key"
    for slot in range(1, 12)
})

away_slots_df = (
    slot_cluster_df[slot_cluster_df["side"].eq("away")]
    .drop(columns="side")
    .rename(columns={"team_id": "away_team_id"})
)
away_slots_df = away_slots_df.rename(columns={
    f"player_{slot:02d}_cluster_key": f"away_player_{slot:02d}_cluster_key"
    for slot in range(1, 12)
})

home_feature_columns = [f"home_player_{slot:02d}_cluster_key" for slot in range(1, 12)]
away_feature_columns = [f"away_player_{slot:02d}_cluster_key" for slot in range(1, 12)]
feature_columns = home_feature_columns + away_feature_columns

match_cluster_dataset_all_df = (
    base_match_df
    .merge(home_slots_df, on=["match_id", "home_team_id"], how="inner")
    .merge(away_slots_df, on=["match_id", "away_team_id"], how="inner")
)

missing_feature_rows = match_cluster_dataset_all_df[feature_columns].isna().any(axis=1).sum()
if missing_feature_rows:
    raise ValueError(f"{missing_feature_rows} matchs ont des clusters manquants dans les 22 slots.")

print("Matchs avec score:", len(base_match_df))
print("Équipes avec 11 titulaires + 11 clusters + 1 gardien:", len(complete_team_lineups_df))
print("Matchs éligibles global:", len(match_cluster_dataset_all_df))
print("Distribution target")
display(match_cluster_dataset_all_df["target_result"].value_counts().sort_index())
display(match_cluster_dataset_all_df[["match_id", "match_date", "home_team_name", "away_team_name", "home_score", "away_score", "target_result"] + feature_columns].head())

Matchs avec score: 4052
Équipes avec 11 titulaires + 11 clusters + 1 gardien: 7471
Matchs éligibles global: 3480
Distribution target


target_result
-1    1069
 0     774
 1    1637
Name: count, dtype: int64

,match_id,match_date,home_team_name,away_team_name,home_score,away_score,target_result,home_player_01_cluster_key,home_player_02_cluster_key,home_player_03_cluster_key,home_player_04_cluster_key,home_player_05_cluster_key,home_player_06_cluster_key,home_player_07_cluster_key,home_player_08_cluster_key,home_player_09_cluster_key,home_player_10_cluster_key,home_player_11_cluster_key,away_player_01_cluster_key,away_player_02_cluster_key,away_player_03_cluster_key,away_player_04_cluster_key,away_player_05_cluster_key,away_player_06_cluster_key,away_player_07_cluster_key,away_player_08_cluster_key,away_player_09_cluster_key,away_player_10_cluster_key,away_player_11_cluster_key
0,3583624,2021-07-07,Fotballklubben Bodø/Glimt,Legia Warszawa,2,3,-1,goalkeeper_0,central_defender_3,central_defender_3,fullback_wingback_0,fullback_wingback_0,central_midfielder_3,central_midfielder_1,central_midfielder_2,forward_2,wide_player_1,forward_2,goalkeeper_1,central_defender_2,central_defender_3,central_defender_3,central_midfielder_1,central_midfielder_3,attacking_midfielder_2,wide_player_0,fullback_wingback_2,wide_player_1,forward_2
1,3583639,2021-07-14,Legia Warszawa,Fotballklubben Bodø/Glimt,2,0,1,goalkeeper_1,central_defender_2,central_defender_3,central_defender_3,central_midfielder_1,central_midfielder_3,attacking_midfielder_2,wide_player_0,fullback_wingback_2,fullback_wingback_2,forward_2,goalkeeper_0,central_defender_2,central_defender_3,central_defender_3,central_midfielder_3,central_midfielder_2,wide_player_1,central_midfielder_1,fullback_wingback_0,attacking_midfielder_1,forward_2
2,3604605,2021-07-17,Club Brugge KV,KRC Genk,3,2,1,goalkeeper_1,central_defender_1,central_defender_3,central_defender_3,central_defender_0,central_defender_2,central_defender_2,defensive_midfielder_2,attacking_midfielder_0,wide_player_0,forward_0,goalkeeper_0,central_defender_2,central_defender_2,fullback_wingback_2,wide_player_1,defensive_midfielder_2,attacking_midfielder_2,wide_player_0,wide_player_0,wide_player_3,forward_0
3,3584558,2021-07-20,Celtic FC,FC Midtjylland,1,1,0,goalkeeper_1,central_defender_2,defensive_midfielder_1,fullback_wingback_2,fullback_wingback_0,defensive_midfielder_0,attacking_midfielder_2,attacking_midfielder_0,defensive_midfielder_2,forward_1,forward_1,goalkeeper_1,central_defender_1,central_defender_3,fullback_wingback_2,central_defender_3,attacking_midfielder_2,defensive_midfielder_1,wide_player_1,wide_player_3,wide_player_0,forward_2
4,3584557,2021-07-20,Rapid Vienna,AC Sparta Prague,2,1,1,goalkeeper_1,central_defender_1,central_defender_2,fullback_wingback_0,fullback_wingback_0,central_midfielder_1,central_midfielder_2,attacking_midfielder_2,wide_player_4,wide_player_2,forward_2,goalkeeper_0,central_defender_2,fullback_wingback_0,fullback_wingback_2,fullback_wingback_0,central_defender_0,central_defender_0,attacking_midfielder_0,wide_player_3,wide_player_1,forward_1


## 8. Ajouter Les Buckets UEFA

Le dataset global conserve tous les matchs éligibles. Le dataset bucket conserve seulement les matchs où les deux équipes sont dans le même intervalle de ranking.

In [85]:
team_rank_features_df = team_rank_mapping_df[[
    "team_id", "rank", "ranking_row", "club_name", "club_key", "total_points", "rank_bucket"
]].copy()

home_rank_df = team_rank_features_df.rename(columns={
    "team_id": "home_team_id",
    "rank": "home_rank",
    "ranking_row": "home_ranking_row",
    "club_name": "home_ranking_club_name",
    "club_key": "home_ranking_club_key",
    "total_points": "home_ranking_points",
    "rank_bucket": "home_rank_bucket",
})

away_rank_df = team_rank_features_df.rename(columns={
    "team_id": "away_team_id",
    "rank": "away_rank",
    "ranking_row": "away_ranking_row",
    "club_name": "away_ranking_club_name",
    "club_key": "away_ranking_club_key",
    "total_points": "away_ranking_points",
    "rank_bucket": "away_rank_bucket",
})

match_cluster_dataset_all_df = (
    match_cluster_dataset_all_df
    .merge(home_rank_df, on="home_team_id", how="left")
    .merge(away_rank_df, on="away_team_id", how="left")
)

same_bucket_mask = (
    match_cluster_dataset_all_df["home_rank_bucket"].notna()
    & match_cluster_dataset_all_df["away_rank_bucket"].notna()
    & (
        match_cluster_dataset_all_df["home_rank_bucket"].astype("Int64")
        == match_cluster_dataset_all_df["away_rank_bucket"].astype("Int64")
    )
)
match_cluster_dataset_same_rank_bucket_df = match_cluster_dataset_all_df[same_bucket_mask].copy()

print("Matchs éligibles global:", len(match_cluster_dataset_all_df))
print("Matchs même bucket UEFA:", len(match_cluster_dataset_same_rank_bucket_df))
print("Distribution bucket")
display(match_cluster_dataset_same_rank_bucket_df["home_rank_bucket"].value_counts().sort_index())
print("Distribution target global")
display(match_cluster_dataset_all_df["target_result"].value_counts().sort_index())
print("Distribution target même bucket")
display(match_cluster_dataset_same_rank_bucket_df["target_result"].value_counts().sort_index())

Matchs éligibles global: 3480
Matchs même bucket UEFA: 1004
Distribution bucket


home_rank_bucket
1.0    555
2.0    242
3.0     73
4.0     70
5.0     64
Name: count, dtype: int64

Distribution target global


target_result
-1    1069
 0     774
 1    1637
Name: count, dtype: int64

Distribution target même bucket


target_result
-1    316
 0    221
 1    467
Name: count, dtype: int64

## 9. Datasets Modèle

On garde les datasets en mémoire pour l'analyse. Les CSV d'analyse ne sont plus écrits : les métriques et conclusions doivent rester visibles dans le notebook.


In [86]:
print("Dataset global:", match_cluster_dataset_all_df.shape)
print("Dataset même bucket UEFA:", match_cluster_dataset_same_rank_bucket_df.shape)
print("Aucun CSV d'analyse n'est écrit dans cette section.")

display(match_cluster_dataset_all_df.head())


Dataset global: (3480, 51)
Dataset même bucket UEFA: (1004, 51)
Aucun CSV d'analyse n'est écrit dans cette section.


,match_id,match_date,competition_id,round,stadium,referee,home_team_id,home_score,home_manager_name,home_formation,away_team_id,away_score,away_manager_name,away_formation,home_team_name,away_team_name,target_result,home_player_01_cluster_key,home_player_02_cluster_key,home_player_03_cluster_key,home_player_04_cluster_key,home_player_05_cluster_key,home_player_06_cluster_key,home_player_07_cluster_key,home_player_08_cluster_key,home_player_09_cluster_key,home_player_10_cluster_key,home_player_11_cluster_key,away_player_01_cluster_key,away_player_02_cluster_key,away_player_03_cluster_key,away_player_04_cluster_key,away_player_05_cluster_key,away_player_06_cluster_key,away_player_07_cluster_key,away_player_08_cluster_key,away_player_09_cluster_key,away_player_10_cluster_key,away_player_11_cluster_key,home_rank,home_ranking_row,home_ranking_club_name,home_ranking_club_key,home_ranking_points,home_rank_bucket,away_rank,away_ranking_row,away_ranking_club_name,away_ranking_club_key,away_ranking_points,away_rank_bucket
0,3583624,2021-07-07,CLQ,First Round 1st leg,Aspmyra Stadion,Juxhin Xhaja,501,2,Kjetil Knutsen,4-3-3 Attacking,255,3,Czeslaw Michniewicz,3-4-2-1,Fotballklubben Bodø/Glimt,Legia Warszawa,-1,goalkeeper_0,central_defender_3,central_defender_3,fullback_wingback_0,fullback_wingback_0,central_midfielder_3,central_midfielder_1,central_midfielder_2,forward_2,wide_player_1,forward_2,goalkeeper_1,central_defender_2,central_defender_3,central_defender_3,central_midfielder_1,central_midfielder_3,attacking_midfielder_2,wide_player_0,fullback_wingback_2,wide_player_1,forward_2,32.0,32.0,Bodø/Glimt,bodo glimt,64.00,2.0,68.0,68.0,Legia Warsaw,legia warsaw,32.50,4.0
1,3583639,2021-07-14,CLQ,First Round 2nd leg,Stadion Miejski im. Marszałka Józefa Piłsudskiego,Aristotelis Diamantopoulos,255,2,Czeslaw Michniewicz,3-4-2-1,501,0,Kjetil Knutsen,3-5-2 flat,Legia Warszawa,Fotballklubben Bodø/Glimt,1,goalkeeper_1,central_defender_2,central_defender_3,central_defender_3,central_midfielder_1,central_midfielder_3,attacking_midfielder_2,wide_player_0,fullback_wingback_2,fullback_wingback_2,forward_2,goalkeeper_0,central_defender_2,central_defender_3,central_defender_3,central_midfielder_3,central_midfielder_2,wide_player_1,central_midfielder_1,fullback_wingback_0,attacking_midfielder_1,forward_2,68.0,68.0,Legia Warsaw,legia warsaw,32.50,4.0,32.0,32.0,Bodø/Glimt,bodo glimt,64.00,2.0
2,3604605,2021-07-17,BESC,Final,Jan Breydelstadion,Jonathan Lardot,2282,3,Philippe Clement,4-3-1-2,1184,2,John van den Brom,4-3-3 Defending,Club Brugge KV,KRC Genk,1,goalkeeper_1,central_defender_1,central_defender_3,central_defender_3,central_defender_0,central_defender_2,central_defender_2,defensive_midfielder_2,attacking_midfielder_0,wide_player_0,forward_0,goalkeeper_0,central_defender_2,central_defender_2,fullback_wingback_2,wide_player_1,defensive_midfielder_2,attacking_midfielder_2,wide_player_0,wide_player_0,wide_player_3,forward_0,23.0,23.0,Club Brugge,club brugge,75.25,2.0,77.0,77.0,Racing Genk,genk,29.00,4.0
3,3584558,2021-07-20,CLQ,Second Round 1st leg,Celtic Park,Sandro Schärer,371,1,Ange Postecoglou,4-2-3-1,865,1,Bo Henriksen,4-3-3 Attacking,Celtic FC,FC Midtjylland,0,goalkeeper_1,central_defender_2,defensive_midfielder_1,fullback_wingback_2,fullback_wingback_0,defensive_midfielder_0,attacking_midfielder_2,attacking_midfielder_0,defensive_midfielder_2,forward_1,forward_1,goalkeeper_1,central_defender_1,central_defender_3,fullback_wingback_2,central_defender_3,attacking_midfielder_2,defensive_midfielder_1,wide_player_1,wide_player_3,wide_player_0,forward_2,58.0,58.0,Celtic,celtic,44.00,3.0,52.0,52.0,FC Midtjylland,midtjylland,48.25,3.0
4,3584557,2021-07-20,CLQ,Second Round 1st leg,Allianz-Stadion,Serdar Gözübüyük,170,2,Dietmar Kühbauer,4-2-3-1,197,1,Pavel Vrba,4-4-1-1,Rapid Vienna,AC Sparta Prague,1,goalkeeper_1,central_defender_1,central_defender_2,fullback_wingback_0,fullback_wingback_0,central_midfielder_1,central_midfielder_2,attacking_midfielder_2,wide

## 10. Fonctions D'Évaluation ML

On utilise un split chronologique : les matchs les plus anciens servent au train et les plus récents au test.

Les métriques train et test sont affichées ensemble pour repérer un éventuel surapprentissage.


In [87]:
from sklearn.dummy import DummyClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.impute import SimpleImputer
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score, balanced_accuracy_score, confusion_matrix, f1_score
from sklearn.model_selection import ParameterGrid
from sklearn.pipeline import Pipeline
from sklearn.preprocessing import OneHotEncoder, StandardScaler


def chronological_train_test_split(frame, test_size=TEST_SIZE):
    ordered_df = frame.sort_values(["match_date", "match_id"]).reset_index(drop=True)
    split_index = int(len(ordered_df) * (1 - test_size))

    if split_index <= 0 or split_index >= len(ordered_df):
        raise ValueError(f"Dataset trop petit pour split train/test: {len(ordered_df)} lignes")

    train_df = ordered_df.iloc[:split_index].copy()
    test_df = ordered_df.iloc[split_index:].copy()
    return train_df, test_df


def compute_classification_metrics(y_true, y_pred):
    return {
        "accuracy": accuracy_score(y_true, y_pred),
        "balanced_accuracy": balanced_accuracy_score(y_true, y_pred),
        "macro_f1": f1_score(y_true, y_pred, average="macro", zero_division=0),
    }


def build_confusion_matrix_df(y_true, y_pred):
    matrix = confusion_matrix(y_true, y_pred, labels=TARGET_LABELS)
    return pd.DataFrame(
        matrix,
        index=[f"actual_{label}" for label in TARGET_LABELS],
        columns=[f"pred_{label}" for label in TARGET_LABELS],
    )


def get_slot_models():
    return {
        "dummy_most_frequent": DummyClassifier(strategy="most_frequent"),
        "logistic_regression": Pipeline([
            ("one_hot", OneHotEncoder(handle_unknown="ignore")),
            ("model", LogisticRegression(
                max_iter=2000,
                class_weight="balanced",
                C=1.0,
                solver="lbfgs",
                random_state=RANDOM_STATE,
            )),
        ]),
        "random_forest": Pipeline([
            ("one_hot", OneHotEncoder(handle_unknown="ignore")),
            ("model", RandomForestClassifier(
                n_estimators=300,
                max_depth=8,
                min_samples_leaf=10,
                class_weight="balanced_subsample",
                random_state=RANDOM_STATE,
                n_jobs=-1,
            )),
        ]),
    }


def evaluate_dataset(dataset_name, dataset_df, input_columns, models):
    train_df, test_df = chronological_train_test_split(dataset_df)
    X_train = train_df[input_columns]
    y_train = train_df["target_result"]
    X_test = test_df[input_columns]
    y_test = test_df["target_result"]

    rows = []
    confusion_tables = {}

    print()
    print(dataset_name)
    print("train:", len(train_df), "test:", len(test_df))
    print("test dates:", test_df["match_date"].min(), "->", test_df["match_date"].max())

    target_distribution_df = pd.DataFrame({
        "train": y_train.value_counts(normalize=True).sort_index(),
        "test": y_test.value_counts(normalize=True).sort_index(),
    }).fillna(0).round(3)
    display(target_distribution_df)

    for model_name, model in models.items():
        model.fit(X_train, y_train)
        y_train_pred = model.predict(X_train)
        y_test_pred = model.predict(X_test)

        train_metrics = compute_classification_metrics(y_train, y_train_pred)
        test_metrics = compute_classification_metrics(y_test, y_test_pred)

        rows.append({
            "dataset": dataset_name,
            "model": model_name,
            "train_rows": len(train_df),
            "test_rows": len(test_df),
            "train_accuracy": train_metrics["accuracy"],
            "train_balanced_accuracy": train_metrics["balanced_accuracy"],
            "train_macro_f1": train_metrics["macro_f1"],
            "test_accuracy": test_metrics["accuracy"],
            "test_balanced_accuracy": test_metrics["balanced_accuracy"],
            "test_macro_f1": test_metrics["macro_f1"],
            "balanced_accuracy_gap": train_metrics["balanced_accuracy"] - test_metrics["balanced_accuracy"],
        })

        confusion_tables[model_name] = build_confusion_matrix_df(y_test, y_test_pred)

    metrics_df = pd.DataFrame(rows).sort_values("test_balanced_accuracy", ascending=False)
    display(metrics_df)

    for model_name, confusion_df in confusion_tables.items():
        print(f"Matrice de confusion test - {model_name}")
        display(confusion_df)

    return metrics_df, confusion_tables


## 11. Baseline Globale

Première lecture : on garde la représentation actuelle avec les 22 `cluster_key` ordonnés par équipe.

Ce bloc sert de référence avant les cross-features.


In [88]:
global_metrics_df, global_confusion_tables = evaluate_dataset(
    "all_matches_slots",
    match_cluster_dataset_all_df,
    feature_columns,
    get_slot_models(),
)



all_matches_slots
train: 2784 test: 696
test dates: 2025-09-18 -> 2026-05-24


,train,test
target_result,,
-1,0.305,0.315
0,0.224,0.217
1,0.471,0.468


,dataset,model,train_rows,test_rows,train_accuracy,train_balanced_accuracy,train_macro_f1,test_accuracy,test_balanced_accuracy,test_macro_f1,balanced_accuracy_gap
2,all_matches_slots,random_forest,2784,696,0.647270,0.643402,0.635709,0.433908,0.399041,0.397427,0.244361
1,all_matches_slots,logistic_regression,2784,696,0.568966,0.575591,0.560816,0.390805,0.369540,0.368282,0.206051
0,all_matches_slots,dummy_most_frequent,2784,696,0.470905,0.333333,0.213431,0.468391,0.333333,0.212655,0.000000


Matrice de confusion test - dummy_most_frequent


,pred_-1,pred_0,pred_1
actual_-1,0,0,219
actual_0,0,0,151
actual_1,0,0,326


Matrice de confusion test - logistic_regression


,pred_-1,pred_0,pred_1
actual_-1,81,51,87
actual_0,48,43,60
actual_1,89,89,148


Matrice de confusion test - random_forest


,pred_-1,pred_0,pred_1
actual_-1,100,29,90
actual_0,61,34,56
actual_1,107,51,168


### Analyse Du Bloc Précédent

Sur le dataset global, on a `2 784` matchs en train et `696` en test.

Le `dummy_most_frequent` prédit toujours la victoire domicile : son accuracy test est `0.468`, mais sa balanced accuracy est `0.333`, donc il ne comprend aucune classe minoritaire.

La LogisticRegression monte à `0.576` de balanced accuracy sur le train mais tombe à `0.370` sur le test. Le gap `0.206` montre déjà du surapprentissage.

Le RandomForest est le meilleur modèle baseline sur le test : balanced accuracy `0.399`, macro F1 `0.397`. Mais il a `0.643` de balanced accuracy sur le train, donc un gap `0.244`. Il apprend bien le train, mais généralise mal.

La matrice de confusion du RandomForest confirme le problème :

- défaites domicile : `100 / 219` bien détectées ;
- nuls : seulement `34 / 151` bien détectés ;
- victoires domicile : `168 / 326` bien détectées.

La classe `nul` est très mal captée. Le modèle confond beaucoup de nuls avec des victoires ou défaites, ce qui est logique avec des features qui décrivent surtout des compositions, pas la probabilité d'un match fermé.


## 12. Baseline Même Bucket UEFA

On refait exactement la même analyse en limitant aux matchs où les deux équipes sont dans le même intervalle UEFA de 20 places.

Le but est de réduire l'écart de niveau brut entre équipes.


In [89]:
same_bucket_metrics_df, same_bucket_confusion_tables = evaluate_dataset(
    "same_rank_bucket_slots",
    match_cluster_dataset_same_rank_bucket_df,
    feature_columns,
    get_slot_models(),
)



same_rank_bucket_slots
train: 803 test: 201
test dates: 2025-09-24 -> 2026-05-24


,train,test
target_result,,
-1,0.313,0.323
0,0.224,0.204
1,0.463,0.473


,dataset,model,train_rows,test_rows,train_accuracy,train_balanced_accuracy,train_macro_f1,test_accuracy,test_balanced_accuracy,test_macro_f1,balanced_accuracy_gap
2,same_rank_bucket_slots,random_forest,803,201,0.687422,0.687927,0.675117,0.398010,0.393035,0.379806,0.294892
1,same_rank_bucket_slots,logistic_regression,803,201,0.724782,0.743832,0.722519,0.363184,0.368237,0.350713,0.375595
0,same_rank_bucket_slots,dummy_most_frequent,803,201,0.463263,0.333333,0.211064,0.472637,0.333333,0.213964,0.000000


Matrice de confusion test - dummy_most_frequent


,pred_-1,pred_0,pred_1
actual_-1,0,0,65
actual_0,0,0,41
actual_1,0,0,95


Matrice de confusion test - logistic_regression


,pred_-1,pred_0,pred_1
actual_-1,15,18,32
actual_0,13,19,9
actual_1,28,28,39


Matrice de confusion test - random_forest


,pred_-1,pred_0,pred_1
actual_-1,18,16,31
actual_0,11,18,12
actual_1,21,30,44


### Analyse Du Bloc Précédent

Sur le dataset `same_rank_bucket`, on a `803` matchs en train et `201` en test. Le filtre réduit l'écart de niveau entre équipes, mais il réduit aussi fortement le volume de données.

Le dummy reste à `0.333` de balanced accuracy, donc il ne sert que de point de référence.

La LogisticRegression surapprend énormément : balanced accuracy train `0.744`, test `0.368`, gap `0.376`. C'est un signal très clair que les 22 slots de joueurs sont trop spécifiques pour seulement `803` matchs de train.

Le RandomForest généralise un peu mieux que la LogisticRegression, mais reste fragile : balanced accuracy train `0.688`, test `0.393`, gap `0.295`.

La matrice de confusion du RandomForest montre :

- défaites domicile : `18 / 65` bien détectées ;
- nuls : `18 / 41` bien détectés ;
- victoires domicile : `44 / 95` bien détectées.

Le filtre same bucket ne suffit donc pas. Le problème principal n'est pas seulement l'écart de niveau entre clubs : c'est surtout la représentation `22 slots`, trop proche d'une composition exacte.


## 13. Lecture De La Baseline

On compare les modèles actuels sans écrire de CSV d'analyse.

La colonne importante pour le surapprentissage est `balanced_accuracy_gap` : plus elle est élevée, plus le modèle réussit sur le train sans généraliser sur le test.


In [90]:
baseline_metrics_df = pd.concat(
    [global_metrics_df, same_bucket_metrics_df],
    ignore_index=True,
).sort_values(["dataset", "test_balanced_accuracy"], ascending=[True, False])

baseline_best_df = (
    baseline_metrics_df
    .sort_values("test_balanced_accuracy", ascending=False)
    .groupby("dataset", as_index=False)
    .head(1)
    .reset_index(drop=True)
)

display(baseline_metrics_df)
print("Meilleur modèle baseline par dataset")
display(baseline_best_df)


,dataset,model,train_rows,test_rows,train_accuracy,train_balanced_accuracy,train_macro_f1,test_accuracy,test_balanced_accuracy,test_macro_f1,balanced_accuracy_gap
0,all_matches_slots,random_forest,2784,696,0.647270,0.643402,0.635709,0.433908,0.399041,0.397427,0.244361
1,all_matches_slots,logistic_regression,2784,696,0.568966,0.575591,0.560816,0.390805,0.369540,0.368282,0.206051
2,all_matches_slots,dummy_most_frequent,2784,696,0.470905,0.333333,0.213431,0.468391,0.333333,0.212655,0.000000
3,same_rank_bucket_slots,random_forest,803,201,0.687422,0.687927,0.675117,0.398010,0.393035,0.379806,0.294892
4,same_rank_bucket_slots,logistic_regression,803,201,0.724782,0.743832,0.722519,0.363184,0.368237,0.350713,0.375595
5,same_rank_bucket_slots,dummy_most_frequent,803,201,0.463263,0.333333,0.211064,0.472637,0.333333,0.213964,0.000000


Meilleur modèle baseline par dataset


,dataset,model,train_rows,test_rows,train_accuracy,train_balanced_accuracy,train_macro_f1,test_accuracy,test_balanced_accuracy,test_macro_f1,balanced_accuracy_gap
0,all_matches_slots,random_forest,2784,696,0.647270,0.643402,0.635709,0.433908,0.399041,0.397427,0.244361
1,same_rank_bucket_slots,random_forest,803,201,0.687422,0.687927,0.675117,0.398010,0.393035,0.379806,0.294892


### Analyse Du Bloc Précédent

La baseline donne un résultat utile mais limité.

Le meilleur modèle global est le RandomForest avec `0.399` de balanced accuracy test. Le meilleur modèle same bucket est aussi le RandomForest avec `0.393` de balanced accuracy test.

Ces scores sont au-dessus du dummy sur les métriques équilibrées, donc il y a un peu de signal dans les clusters joueurs. Mais le gap train-test est trop élevé : `0.244` en global et `0.295` en same bucket pour le RandomForest.

Conclusion : les clusters individuels sont probablement intéressants, mais la forme `home_player_01`, `home_player_02`, etc. est trop rigide. Deux compos proches peuvent devenir deux lignes très différentes juste parce que les joueurs sont ordonnés différemment.


## 14. Cross-Features De Composition

On change la représentation du match.

Au lieu de donner au modèle 22 slots ordonnés, on décrit chaque équipe par le nombre de joueurs de chaque `cluster_key`, puis on ajoute les différences home-away.

On sépare volontairement les features en trois groupes :

- `styles_only` : uniquement la composition de styles ;
- `ranking_only` : uniquement la différence de niveau UEFA ;
- `styles_plus_ranking` : les deux ensemble.


In [91]:
def build_style_cross_feature_dataset(source_df, cluster_key_values):
    home_slots_df = source_df[home_feature_columns]
    away_slots_df = source_df[away_feature_columns]

    home_count_core_df = pd.DataFrame(
        {cluster_key: home_slots_df.eq(cluster_key).sum(axis=1) for cluster_key in cluster_key_values},
        index=source_df.index,
    )
    away_count_core_df = pd.DataFrame(
        {cluster_key: away_slots_df.eq(cluster_key).sum(axis=1) for cluster_key in cluster_key_values},
        index=source_df.index,
    )

    home_count_df = home_count_core_df.add_prefix("home_count__")
    away_count_df = away_count_core_df.add_prefix("away_count__")
    delta_count_df = (home_count_core_df - away_count_core_df).add_prefix("delta_count__")

    style_features_df = pd.concat([home_count_df, away_count_df, delta_count_df], axis=1)

    # Similarité globale entre les deux compositions de styles.
    style_features_df["style_overlap_count"] = pd.DataFrame(
        np.minimum(home_count_core_df.to_numpy(), away_count_core_df.to_numpy()),
        index=source_df.index,
    ).sum(axis=1)

    # Diversité de styles dans chaque onze titulaire.
    style_features_df["home_style_diversity"] = home_count_core_df.gt(0).sum(axis=1)
    style_features_df["away_style_diversity"] = away_count_core_df.gt(0).sum(axis=1)

    ranking_features_df = pd.DataFrame({
        "rank_diff_home_advantage": source_df["away_rank"] - source_df["home_rank"],
        "points_diff_home_advantage": source_df["home_ranking_points"] - source_df["away_ranking_points"],
        "bucket_diff_home_advantage": source_df["away_rank_bucket"] - source_df["home_rank_bucket"],
    }, index=source_df.index)

    base_columns = [
        "match_id", "match_date", "home_team_name", "away_team_name",
        "home_score", "away_score", "target_result",
    ]
    available_base_columns = [column for column in base_columns if column in source_df.columns]

    model_df = pd.concat(
        [
            source_df[available_base_columns].reset_index(drop=True),
            style_features_df.reset_index(drop=True),
            ranking_features_df.reset_index(drop=True),
        ],
        axis=1,
    )
    return model_df, list(style_features_df.columns), list(ranking_features_df.columns)


cluster_key_values = sorted(clusters_df["cluster_key"].unique())

global_cross_feature_df, style_feature_columns, ranking_feature_columns = build_style_cross_feature_dataset(
    match_cluster_dataset_all_df,
    cluster_key_values,
)
same_bucket_cross_feature_df, _, _ = build_style_cross_feature_dataset(
    match_cluster_dataset_same_rank_bucket_df,
    cluster_key_values,
)

cross_feature_groups = {
    "styles_only": style_feature_columns,
    "ranking_only": ranking_feature_columns,
    "styles_plus_ranking": style_feature_columns + ranking_feature_columns,
}

print("Features styles:", len(style_feature_columns))
print("Features ranking:", len(ranking_feature_columns))
print("Global:", global_cross_feature_df.shape)
print("Même bucket:", same_bucket_cross_feature_df.shape)
display(global_cross_feature_df.head())


Features styles: 90
Features ranking: 3
Global: (3480, 100)
Même bucket: (1004, 100)


,match_id,match_date,home_team_name,away_team_name,home_score,away_score,target_result,home_count__attacking_midfielder_0,home_count__attacking_midfielder_1,home_count__attacking_midfielder_2,home_count__central_defender_0,home_count__central_defender_1,home_count__central_defender_2,home_count__central_defender_3,home_count__central_midfielder_0,home_count__central_midfielder_1,home_count__central_midfielder_2,home_count__central_midfielder_3,home_count__defensive_midfielder_0,home_count__defensive_midfielder_1,home_count__defensive_midfielder_2,home_count__defensive_midfielder_3,home_count__defensive_midfielder_4,home_count__forward_0,home_count__forward_1,home_count__forward_2,home_count__fullback_wingback_0,home_count__fullback_wingback_1,home_count__fullback_wingback_2,home_count__goalkeeper_0,home_count__goalkeeper_1,home_count__wide_player_0,home_count__wide_player_1,home_count__wide_player_2,home_count__wide_player_3,home_count__wide_player_4,away_count__attacking_midfielder_0,away_count__attacking_midfielder_1,away_count__attacking_midfielder_2,away_count__central_defender_0,away_count__central_defender_1,away_count__central_defender_2,away_count__central_defender_3,away_count__central_midfielder_0,away_count__central_midfielder_1,away_count__central_midfielder_2,away_count__central_midfielder_3,away_count__defensive_midfielder_0,away_count__defensive_midfielder_1,away_count__defensive_midfielder_2,away_count__defensive_midfielder_3,away_count__defensive_midfielder_4,away_count__forward_0,away_count__forward_1,away_count__forward_2,away_count__fullback_wingback_0,away_count__fullback_wingback_1,away_count__fullback_wingback_2,away_count__goalkeeper_0,away_count__goalkeeper_1,away_count__wide_player_0,away_count__wide_player_1,away_count__wide_player_2,away_count__wide_player_3,away_count__wide_player_4,delta_count__attacking_midfielder_0,delta_count__attacking_midfielder_1,delta_count__attacking_midfielder_2,delta_count__central_defender_0,delta_count__central_defender_1,delta_count__central_defender_2,delta_count__central_defender_3,delta_count__central_midfielder_0,delta_count__central_midfielder_1,delta_count__central_midfielder_2,delta_count__central_midfielder_3,delta_count__defensive_midfielder_0,delta_count__defensive_midfielder_1,delta_count__defensive_midfielder_2,delta_count__defensive_midfielder_3,delta_count__defensive_midfielder_4,delta_count__forward_0,delta_count__forward_1,delta_count__forward_2,delta_count__fullback_wingback_0,delta_count__fullback_wingback_1,delta_count__fullback_wingback_2,delta_count__goalkeeper_0,delta_count__goalkeeper_1,delta_count__wide_player_0,delta_count__wide_player_1,delta_count__wide_player_2,delta_count__wide_player_3,delta_count__wide_player_4,style_overlap_count,home_style_diversity,away_style_diversity,rank_diff_home_advantage,points_diff_home_advantage,bucket_diff_home_advantage
0,3583624,2021-07-07,Fotballklubben Bodø/Glimt,Legia Warszawa,2,3,-1,0,0,0,0,0,0,2,0,1,1,1,0,0,0,0,0,0,0,2,2,0,0,1,0,0,1,0,0,0,0,0,1,0,0,1,2,0,1,0,1,0,0,0,0,0,0,0,1,0,0,1,0,1,1,1,0,0,0,0,0,-1,0,0,-1,0,0,0,1,0,0,0,0,0,0,0,0,1,2,0,-1,1,-1,-1,0,0,0,0,6,8,10,36.0,31.50,2.0
1,3583639,2021-07-14,Legia Warszawa,Fotballklubben Bodø/Glimt,2,0,1,0,0,1,0,0,1,2,0,1,0,1,0,0,0,0,0,0,0,1,0,0,2,0,1,1,0,0,0,0,0,1,0,0,0,1,2,0,1,1,1,0,0,0,0,0,0,0,1,1,0,0,1,0,0,1,0,0,0,0,-1,1,0,0,0,0,0,0,-1,0,0,0,0,0,0,0,0,0,-1,0,2,-1,1,1,-1,0,0,0,6,9,10,-36.0,-31.50,-2.0
2,3604605,2021-07-17,Club Brugge KV,KRC Genk,3,2,1,1,0,0,1,1,2,2,0,0,0,0,0,0,1,0,0,1,0,0,0,0,0,0,1,1,0,0,0,0,0,0,1,0,0,2,0,0,0,0,0,0,0,1,0,0,1,0,0,0,0,1,1,0,2,1,0,1,0,1,0,-1,1,1,0,2,0,0,0,0,0,0,0,0,0,0,0,0,0,0,-1,-1,1,-1,-1,0,-1,0,5,9,9,54.0,46.25,2.0
3,3584558,2021-07-20,Celtic FC,FC Midtjylland,1,1,0,1,0,1,0,0,1,0,0,0,0,0,1,1,1,0,0,0,2,0,1,0,1,0,1,0,0,0,0,0,0,0,1,0,1,0,2,0,0,0,0,0,1,0,0,0,0,0,1,0,0,1,0,1,1,1,0,1,0,1,0,0,0,-1,1,-2,0,0,0,0,1,0,1,0,0,0,2,-1,1,0,0,0,0,-1,-1,0,-1,0,4,10,10,-6.0,-4.25,0.0
4,3584557,2021-07-20,Rapid Vienna,AC Sparta Prague,2,1,1,0,0

### Analyse Du Bloc Précédent

Le bloc crée `90` features de styles et `3` features de ranking.

Cette séparation est importante : sans elle, on ne sait pas si le modèle apprend vraiment les compositions ou s'il exploite surtout le niveau UEFA des clubs.

La suite compare donc trois hypothèses : les styles seuls, le ranking seul, puis les deux ensemble.


## 15. Validation Temporelle Et Régularisation

On retravaille le surapprentissage proprement.

Le test final reste intouché. Les hyperparamètres sont choisis dans le train avec plusieurs folds temporels, puis le meilleur modèle est seulement évalué sur le test.

On compare aussi LogisticRegression et RandomForest pour distinguer un problème de features d'un problème de modèle trop flexible.


In [92]:
def make_numeric_logistic_pipeline(params):
    return Pipeline([
        ("imputer", SimpleImputer(strategy="median")),
        ("scaler", StandardScaler()),
        ("model", LogisticRegression(
            **params,
            max_iter=3000,
            class_weight="balanced",
            solver="lbfgs",
            random_state=RANDOM_STATE,
        )),
    ])


def make_regularized_random_forest_pipeline(params):
    return Pipeline([
        ("imputer", SimpleImputer(strategy="median")),
        ("model", RandomForestClassifier(
            **params,
            class_weight="balanced_subsample",
            random_state=RANDOM_STATE,
            n_jobs=-1,
        )),
    ])


def chronological_validation_splits(frame, n_splits=3):
    ordered_df = frame.sort_values(["match_date", "match_id"]).reset_index(drop=True)
    block_size = len(ordered_df) // (n_splits + 1)

    splits = []
    for fold_index in range(n_splits):
        train_end = block_size * (fold_index + 1)
        validation_end = block_size * (fold_index + 2)
        train_fold_df = ordered_df.iloc[:train_end].copy()
        validation_fold_df = ordered_df.iloc[train_end:validation_end].copy()
        splits.append((fold_index + 1, train_fold_df, validation_fold_df))

    return splits


def get_candidate_models():
    candidates = []

    for params in ParameterGrid({"C": [0.1, 1.0]}):
        candidates.append(("logistic_regression", params, make_numeric_logistic_pipeline))

    for params in ParameterGrid({
        "n_estimators": [150],
        "max_depth": [3, 5, 8],
        "min_samples_leaf": [40, 80],
        "max_features": ["sqrt"],
    }):
        candidates.append(("random_forest", params, make_regularized_random_forest_pipeline))

    return candidates


def score_candidate_on_validation_folds(train_df, input_columns, model_name, params, model_factory):
    fold_rows = []

    for fold_index, train_fold_df, validation_fold_df in chronological_validation_splits(train_df):
        model = model_factory(params)
        model.fit(train_fold_df[input_columns], train_fold_df["target_result"])

        train_pred = model.predict(train_fold_df[input_columns])
        validation_pred = model.predict(validation_fold_df[input_columns])
        train_metrics = compute_classification_metrics(train_fold_df["target_result"], train_pred)
        validation_metrics = compute_classification_metrics(validation_fold_df["target_result"], validation_pred)

        fold_rows.append({
            "fold": fold_index,
            "model": model_name,
            "params": params,
            "train_balanced_accuracy": train_metrics["balanced_accuracy"],
            "validation_balanced_accuracy": validation_metrics["balanced_accuracy"],
            "validation_macro_f1": validation_metrics["macro_f1"],
            "validation_gap": train_metrics["balanced_accuracy"] - validation_metrics["balanced_accuracy"],
        })

    return pd.DataFrame(fold_rows)


def tune_models_with_temporal_validation(dataset_name, feature_set_name, dataset_df, input_columns):
    train_df, test_df = chronological_train_test_split(dataset_df)

    tuning_tables = []
    for model_name, params, model_factory in get_candidate_models():
        fold_scores_df = score_candidate_on_validation_folds(
            train_df,
            input_columns,
            model_name,
            params,
            model_factory,
        )
        tuning_tables.append(fold_scores_df)

    tuning_df = pd.concat(tuning_tables, ignore_index=True)
    tuning_summary_df = (
        tuning_df
        .assign(params_text=lambda frame: frame["params"].map(str))
        .groupby(["model", "params_text"], as_index=False)
        .agg(
            mean_validation_balanced_accuracy=("validation_balanced_accuracy", "mean"),
            mean_validation_macro_f1=("validation_macro_f1", "mean"),
            mean_validation_gap=("validation_gap", "mean"),
            params=("params", "first"),
        )
        .sort_values(
            ["mean_validation_balanced_accuracy", "mean_validation_macro_f1", "mean_validation_gap"],
            ascending=[False, False, True],
        )
    )

    best_row = tuning_summary_df.iloc[0]
    best_model_name = best_row["model"]
    best_params = best_row["params"]
    best_factory = make_numeric_logistic_pipeline if best_model_name == "logistic_regression" else make_regularized_random_forest_pipeline

    final_model = best_factory(best_params)
    final_model.fit(train_df[input_columns], train_df["target_result"])

    train_pred = final_model.predict(train_df[input_columns])
    test_pred = final_model.predict(test_df[input_columns])
    train_metrics = compute_classification_metrics(train_df["target_result"], train_pred)
    test_metrics = compute_classification_metrics(test_df["target_result"], test_pred)

    final_metrics_df = pd.DataFrame([{
        "dataset": dataset_name,
        "feature_set": feature_set_name,
        "model": best_model_name,
        "train_rows": len(train_df),
        "test_rows": len(test_df),
        "train_balanced_accuracy": train_metrics["balanced_accuracy"],
        "test_balanced_accuracy": test_metrics["balanced_accuracy"],
        "test_macro_f1": test_metrics["macro_f1"],
        "balanced_accuracy_gap": train_metrics["balanced_accuracy"] - test_metrics["balanced_accuracy"],
        "validation_balanced_accuracy": best_row["mean_validation_balanced_accuracy"],
        "validation_gap": best_row["mean_validation_gap"],
        "best_params": best_params,
    }])

    confusion_df = build_confusion_matrix_df(test_df["target_result"], test_pred)
    return final_metrics_df, confusion_df, tuning_summary_df


def run_regularized_feature_experiment(dataset_name, dataset_df):
    metrics_tables = []
    confusion_tables = {}
    tuning_tables = {}

    for feature_set_name, input_columns in cross_feature_groups.items():
        metrics_df, confusion_df, tuning_df = tune_models_with_temporal_validation(
            dataset_name,
            feature_set_name,
            dataset_df,
            input_columns,
        )
        metrics_tables.append(metrics_df)
        confusion_tables[feature_set_name] = confusion_df
        tuning_tables[feature_set_name] = tuning_df

    metrics_df = pd.concat(metrics_tables, ignore_index=True).sort_values(
        "test_balanced_accuracy",
        ascending=False,
    )

    print(dataset_name)
    display(metrics_df)

    best_feature_set = metrics_df.iloc[0]["feature_set"]
    print("Meilleur feature set:", best_feature_set)
    print("Matrice de confusion test - meilleur modèle régularisé")
    display(confusion_tables[best_feature_set])

    return metrics_df, confusion_tables, tuning_tables


global_regularized_metrics_df, global_regularized_confusions, global_regularized_tuning = run_regularized_feature_experiment(
    "all_matches",
    global_cross_feature_df,
)

same_bucket_regularized_metrics_df, same_bucket_regularized_confusions, same_bucket_regularized_tuning = run_regularized_feature_experiment(
    "same_rank_bucket",
    same_bucket_cross_feature_df,
)


all_matches


,dataset,feature_set,model,train_rows,test_rows,train_balanced_accuracy,test_balanced_accuracy,test_macro_f1,balanced_accuracy_gap,validation_balanced_accuracy,validation_gap,best_params
2,all_matches,styles_plus_ranking,random_forest,2784,696,0.472105,0.434550,0.420662,0.037555,0.424236,0.056157,"{'max_depth': 8, 'max_features': 'sqrt', 'min_..."
1,all_matches,ranking_only,logistic_regression,2784,696,0.430986,0.428878,0.426948,0.002108,0.441906,0.007626,{'C': 1.0}
0,all_matches,styles_only,random_forest,2784,696,0.502182,0.373898,0.371377,0.128284,0.378993,0.128753,"{'max_depth': 5, 'max_features': 'sqrt', 'min_..."


Meilleur feature set: styles_plus_ranking
Matrice de confusion test - meilleur modèle régularisé


,pred_-1,pred_0,pred_1
actual_-1,132,22,65
actual_0,67,22,62
actual_1,113,32,181


same_rank_bucket


,dataset,feature_set,model,train_rows,test_rows,train_balanced_accuracy,test_balanced_accuracy,test_macro_f1,balanced_accuracy_gap,validation_balanced_accuracy,validation_gap,best_params
1,same_rank_bucket,ranking_only,logistic_regression,803,201,0.429553,0.450123,0.448987,-0.020570,0.445082,0.003439,{'C': 0.1}
2,same_rank_bucket,styles_plus_ranking,random_forest,803,201,0.521802,0.428064,0.418388,0.093739,0.401502,0.130705,"{'max_depth': 3, 'max_features': 'sqrt', 'min_..."
0,same_rank_bucket,styles_only,random_forest,803,201,0.535951,0.359508,0.357563,0.176443,0.365419,0.164875,"{'max_depth': 3, 'max_features': 'sqrt', 'min_..."


Meilleur feature set: ranking_only
Matrice de confusion test - meilleur modèle régularisé


,pred_-1,pred_0,pred_1
actual_-1,30,18,17
actual_0,7,14,20
actual_1,22,21,52


### Analyse Du Bloc Précédent

La validation temporelle régularisée change beaucoup la lecture.

Sur le dataset global, le meilleur modèle est `styles_plus_ranking` avec un RandomForest régularisé : `max_depth=8`, `min_samples_leaf=80`, `n_estimators=150`. Il obtient `0.435` de balanced accuracy test, `0.421` de macro F1 et un gap train-test de seulement `0.037`.

C'est nettement plus sain que la baseline `22 slots`, dont le RandomForest avait `0.399` de balanced accuracy test et un gap `0.244`.

Mais l'ablation est essentielle :

- `ranking_only` obtient déjà `0.429` avec un gap quasi nul `0.002` ;
- `styles_only` tombe à `0.374` avec un gap `0.128` ;
- `styles_plus_ranking` monte seulement à `0.435`.

Donc, en global, les styles ajoutent un petit signal au ranking, mais ce signal est faible.

Sur le dataset `same_rank_bucket`, le résultat est encore plus clair :

- `ranking_only` est le meilleur modèle avec `0.450` de balanced accuracy test et un gap `-0.021` ;
- `styles_plus_ranking` descend à `0.428` avec un gap `0.094` ;
- `styles_only` descend à `0.360` avec un gap `0.176`.

Cela veut dire que même dans un bucket UEFA identique, la différence fine de ranking/points reste plus prédictive que les compositions de styles actuelles.

Les matrices de confusion confirment aussi que le nul reste le gros point faible. En global, le meilleur modèle détecte seulement `22 / 151` nuls. En same bucket, le meilleur modèle détecte `14 / 41` nuls. Le modèle apprend surtout à différencier victoire domicile vs victoire extérieur, mais pas encore les matchs équilibrés qui finissent nuls.


## 16. Comparaison Et Diagnostic Data Scientist

On compare maintenant la baseline `22 slots` au meilleur modèle régularisé issu des cross-features.

L'objectif n'est pas seulement d'augmenter le score test : on veut aussi diminuer le gap train-test.


In [93]:
regularized_metrics_df = pd.concat(
    [global_regularized_metrics_df, same_bucket_regularized_metrics_df],
    ignore_index=True,
)

baseline_tree_metrics_df = baseline_metrics_df[
    baseline_metrics_df["model"].eq("random_forest")
].copy()
baseline_tree_metrics_df["feature_set"] = "22_slots"
baseline_tree_metrics_df["validation_balanced_accuracy"] = np.nan
baseline_tree_metrics_df["validation_gap"] = np.nan
baseline_tree_metrics_df["best_params"] = "baseline_not_tuned_with_multifold_cv"
baseline_tree_metrics_df = baseline_tree_metrics_df.rename(columns={
    "test_macro_f1": "test_macro_f1",
})

comparison_columns = [
    "dataset", "feature_set", "model", "train_rows", "test_rows",
    "train_balanced_accuracy", "test_balanced_accuracy", "test_macro_f1",
    "balanced_accuracy_gap", "validation_balanced_accuracy", "validation_gap", "best_params",
]

model_comparison_df = pd.concat(
    [baseline_tree_metrics_df[comparison_columns], regularized_metrics_df[comparison_columns]],
    ignore_index=True,
).sort_values(["dataset", "test_balanced_accuracy"], ascending=[True, False])

display(model_comparison_df)

print("Meilleur modèle par dataset")
best_model_by_dataset_df = (
    model_comparison_df
    .sort_values("test_balanced_accuracy", ascending=False)
    .groupby("dataset", as_index=False)
    .head(1)
    .reset_index(drop=True)
)
display(best_model_by_dataset_df)


,dataset,feature_set,model,train_rows,test_rows,train_balanced_accuracy,test_balanced_accuracy,test_macro_f1,balanced_accuracy_gap,validation_balanced_accuracy,validation_gap,best_params
2,all_matches,styles_plus_ranking,random_forest,2784,696,0.472105,0.434550,0.420662,0.037555,0.424236,0.056157,"{'max_depth': 8, 'max_features': 'sqrt', 'min_..."
3,all_matches,ranking_only,logistic_regression,2784,696,0.430986,0.428878,0.426948,0.002108,0.441906,0.007626,{'C': 1.0}
4,all_matches,styles_only,random_forest,2784,696,0.502182,0.373898,0.371377,0.128284,0.378993,0.128753,"{'max_depth': 5, 'max_features': 'sqrt', 'min_..."
0,all_matches_slots,22_slots,random_forest,2784,696,0.643402,0.399041,0.397427,0.244361,NaN,NaN,baseline_not_tuned_with_multifold_cv
5,same_rank_bucket,ranking_only,logistic_regression,803,201,0.429553,0.450123,0.448987,-0.020570,0.445082,0.003439,{'C': 0.1}
6,same_rank_bucket,styles_plus_ranking,random_forest,803,201,0.521802,0.428064,0.418388,0.093739,0.401502,0.130705,"{'max_depth': 3, 'max_features': 'sqrt', 'min_..."
7,same_rank_bucket,styles_only,random_forest,803,201,0.535951,0.359508,0.357563,0.176443,0.365419,0.164875,"{'max_depth': 3, 'max_features': 'sqrt', 'min_..."
1,same_rank_bucket_slots,22_slots,random_forest,803,201,0.687927,0.393035,0.379806,0.294892,NaN,NaN,baseline_not_tuned_with_multifold_cv


Meilleur modèle par dataset


,dataset,feature_set,model,train_rows,test_rows,train_balanced_accuracy,test_balanced_accuracy,test_macro_f1,balanced_accuracy_gap,validation_balanced_accuracy,validation_gap,best_params
0,same_rank_bucket,ranking_only,logistic_regression,803,201,0.429553,0.450123,0.448987,-0.020570,0.445082,0.003439,{'C': 0.1}
1,all_matches,styles_plus_ranking,random_forest,2784,696,0.472105,0.434550,0.420662,0.037555,0.424236,0.056157,"{'max_depth': 8, 'max_features': 'sqrt', 'min_..."
2,all_matches_slots,22_slots,random_forest,2784,696,0.643402,0.399041,0.397427,0.244361,NaN,NaN,baseline_not_tuned_with_multifold_cv
3,same_rank_bucket_slots,22_slots,random_forest,803,201,0.687927,0.393035,0.379806,0.294892,NaN,NaN,baseline_not_tuned_with_multifold_cv


### Conclusion Des Résultats

Le diagnostic est maintenant plus propre : le surapprentissage ne vient pas seulement du RandomForest. Il vient surtout de la représentation actuelle des styles.

Les `22 slots` sont trop spécifiques. Les cross-features régularisées sont meilleures, mais `styles_only` reste faible : `0.374` en global et `0.360` en same bucket. Cela indique que nos profils de joueurs actuels ne donnent pas encore une mesure stable de la force collective.

Le ranking est aujourd'hui le signal le plus solide : `0.429` en global et `0.450` en same bucket avec très peu de gap train-test. C'est un bon baseline, mais ce n'est pas notre objectif final.

La bonne lecture est donc :

- oui, les styles contiennent peut-être un peu d'information ;
- non, ils ne sont pas encore assez robustes seuls ;
- le modèle risque encore d'apprendre des artefacts de composition si on lui donne trop de colonnes de styles détaillées ;
- pour mesurer une vraie force de composition, il faut compresser les styles.

Prochaine direction data scientist : passer des `90` features de comptes détaillés à une représentation plus stable, par exemple :

- agrégats par rôle : défense, milieux, côtés, attaque, gardien ;
- moyennes des dimensions de style des joueurs plutôt que compte exact des clusters ;
- clustering de compositions collectives, puis analyse `style collectif A vs style collectif B` ;
- modèle `ranking_only` comme baseline, puis mesure du gain marginal réel des styles.

Pour les nuls, il faudra probablement une approche séparée : modèle `draw vs not_draw`, ou prédiction du `goal_diff` avant conversion en `-1 / 0 / 1`.
